core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,539 @@
+		endpoints:                           comm.Shuffle(endpoints),
+		blockReceiverIndex:                  -1,
+		headerReceivers:                     make(map[string]*bftHeaderReceiver),
+	}
+
+	bftLogger.Infof("[%s] Created BFT Delivery Client", chainID)
+	return c
+}
+
+func (c *bftDeliveryClient) Recv() (response *orderer.DeliverResponse, err error) {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	c.startOnce.Do(func() {
+		var num uint64
+		num, err = c.ledgerInfoProvider.LedgerHeight()
+		if err != nil {
+			return
+		}
+		c.nextBlockNumber = num
+		c.lastBlockTime = time.Now()
+		bftLogger.Debugf("[%s] Starting monitor routine; Initial ledger height: %d", c.chainID, num)
+		go c.monitor()
+	})
+	// can only happen once, after first invocation
+	if err != nil {
+		bftLogger.Debugf("[%s] Exit: error=%v", c.chainID, err)
+		return nil, errors.New("cannot access ledger height")
+	}
+
+	var numEP int
+	var numRetries int
+	var stopRetries = time.Now().Add(c.reconnectTotalTimeThreshold)
+
+	for !c.shouldStop() {
+		if numEP, err = c.assignReceivers(); err != nil {
+			bftLogger.Debugf("[%s] Exit: error=%s", c.chainID, err)
+			return nil, err
+		}
+
+		c.launchHeaderReceivers()
+
+		response, err = c.receiveBlock()
+		if err == nil {
+			bftLogger.Debugf("[%s] Exit: response=%v", c.chainID, response)
+			return response, nil // the normal return path
+		}
+
+		if stopRetries.Before(time.Now()) {
+			bftLogger.Debugf("[%s] Exit: reconnectTotalTimeThreshold: %s, expired; error: %s",
+				c.chainID, c.reconnectTotalTimeThreshold, errClientReconnectTimeout)
+			return nil, errClientReconnectTimeout
+		}
+
+		c.closeBlockReceiver()
+		numRetries++
+		if numRetries%numEP == 0 { //double the back-off delay on every round of attempts.
+			dur := backOffDuration(2.0, uint(numRetries/numEP), bftMinBackoffDelay, bftMaxBackoffDelay)
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint in: %s", c.chainID, err, dur)
+			backOffSleep(dur, c.stopChan)
+		} else {
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint now", c.chainID, err)
+		}
+	}
+
+	bftLogger.Debugf("[%s] Exit: %s", c.chainID, errClientClosing.Error())
+	return nil, errClientClosing
+}
+
+func backOffDuration(base float64, exponent uint, minDur, maxDur time.Duration) (backOffDur time.Duration) {
+	if base < 1.0 {
+		base = 1.0
+	}
+	if minDur <= 0 {
+		minDur = time.Duration(1)
+	}
+	if maxDur < minDur {
+		maxDur = minDur
+	}
+
+	fDurNano := math.Pow(base, float64(exponent)) * float64(minDur.Nanoseconds())
+	if math.IsInf(fDurNano, 0) || math.IsNaN(fDurNano) || fDurNano > float64(maxDur.Nanoseconds()) {
+		backOffDur = maxDur
+	} else {
+		backOffDur = time.Duration(int64(fDurNano))
+	}
+
+	return backOffDur
+}
+
+func backOffSleep(backOffDur time.Duration, stopChan <-chan struct{}) {
+	backOffTimer := time.NewTimer(backOffDur)
+	select {
+	case <-backOffTimer.C:
```

Comment:
you don't need a timer here. you can just use time.After

Reply:
done